packages/api/src/car.js

```diff
@@ -227,17 +227,37 @@ async function backup (blob, rootCid, userId, env, structure = 'Unknown') {
     return undefined
   }
 
-  const data = await blob.arrayBuffer()
-  const dataHash = await sha256.digest(new Uint8Array(data))
-  const keyStr = `raw/${rootCid.toString()}/${userId}/${toString(dataHash.bytes, 'base32')}.car`
+  const data = new Uint8Array(await blob.arrayBuffer())
+  const multihash = await sha256.digest(data)
+  const keyStr = `raw/${rootCid.toString()}/${userId}/${toString(multihash.bytes, 'base32')}.car`
+  // strip the multihash varint prefix to get the raw sha256 digest for aws upload integrity check
+  const rawSha256 = multihash.bytes.subarray(2)
+  const awsChecksum = toString(rawSha256, 'base64pad')
+
+  // see: https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/modules/putobjectrequest.html
   const cmdParams = {
     Bucket: env.s3BucketName,
     Key: keyStr,
-    Body: blob,
-    Metadata: { structure }
+    Body: data,
+    Metadata: { structure },
+    // ChecksumSHA256 specifies the base64-encoded, 256-bit SHA-256 digest of the object, used as a data integrity check to verify that the data received is the same data that was originally sent.
+    // see: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#AmazonS3-PutObject-request-header-ChecksumSHA256
+    ChecksumSHA256: awsChecksum
   }
 
-  await env.s3Client.send(new PutObjectCommand(cmdParams))
+  try {
+    await env.s3Client.send(new PutObjectCommand(cmdParams))
+  } catch (err) {
+    if (err.name === 'BadDigest') {
+      // s3 returns a 400 Bad Request `BadDigest` error if the hash does not match their calculation.
+      // see: https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#RESTErrorResponses
+      // see: https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/index.html#troubleshooting
+      console.log('BadDigest: sha256 of data recieved did not match what we sent. Maybe bits flipped in transit. Retrying once.')
```

Comment:
```suggestion
```

can be removed?

Reply:
This was intentional, as I'd be curious to see how many times this appears in the logs, but I could be pursuaded to remove it as I can't say I'd be checking the logs for it (if we had any)

Reply:
We need https://github.com/web3-storage/web3.storage/issues/922 for this. With `console.log` we won't be able to see any logs

Reply:
@vasco-santos want me to remove it? or leave it in ahead of #922

Reply:
can leave, #922 already has this linked, so we just need to remember to change it